Encrypt network communication in JGroups

pruivo opened the pull request:
Fixes #25702

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

Draft PR because I would like to receive feedback from the following
* Is it okay for these options to live in the `cache` category?
* Where to document how to configure it with the Keycloak Operator?

pruivo committed:
Encrypt network communication in JGroups

Fixes #25702

Signed-off-by: Pedro Ruivo <pruivo@redhat.com>
Co-authored-by: Alexander Schwartz <aschwart@redhat.com>

pruivo commented:
I tested with the current Keycloak Operator code locally using `kind` and with the following snipped
```yaml
spec:
...
  additionalOptions:
    - name: jgroups-tls-enabled
      value: "true"
    - name: jgroups-tls-protocol
      value: "TLSv1.3"
    - name: jgroups-tls-keystore-file
      value: /etc/jgroups-tls/keystore.p12
    - name: jgroups-tls-keystore-password
      secret:
        name: keycloak-jgroups-tls-secret
        key: keystorePassword
    - name: jgroups-tls-keystore-type
      value: "pkcs12"
    - name: jgroups-tls-keystore-alias
      value: "site1"
    - name: jgroups-tls-truststore-file
      value: /etc/jgroups-tls/truststore.p12
    - name: jgroups-tls-truststore-password
      secret:
        name: keycloak-jgroups-tls-secret
        key: truststorePassword
    - name: jgroups-tls-keystore-type
      value: "pkcs12"
    - name: jgroups-tls-client-auth
      value: "need"
  instances: 2  
  unsupported:
    podTemplate:
      spec:
        containers:
          - volumeMounts:
              - name: jgroups-tls-volume
                mountPath: /etc/jgroups-tls/
                readOnly: true
        volumes:
          - name: jgroups-tls-volume
            secret:
              secretName: keycloak-jgroups-tls-secret
```
The secret was created using 
```
kubectl create secret generic keycloak-jgroups-tls-secret \
    --from-file=keystore.p12="./site1-keystore.p12" \
    --from-literal=keystorePassword=*** \
    --from-file=truststore.p12="./ca-keystore.p12" \
    --from-literal=truststorePassword=*** \
```

It is probably worth to write these example in some documentation page :+1:

ahus1 commented on quarkus/config-api/src/main/java/org/keycloak/config/CachingOptions.java:
```diff
@@ -40,4 +53,83 @@ public enum Stack {
+            .buildTime(true)
+            .build();
+
+    public static final Option<String> JGROUPS_TLS_PROVIDER = new OptionBuilder<>(JGROUPS_TLS_PROVIDER_PROPERTY, String.class)
```
Do you know that people actually do / need to do this in their environment? If it is not a common thing to do, I'd suggest removing this parameter.

pruivo replied:
We don't allow custom providers in Infinispan Server but we have the option to use the JDK provider or the OpenSSL provider (via WildFly-OpenSSL). I'm going to remove this option.

ahus1 commented on quarkus/config-api/src/main/java/org/keycloak/config/CachingOptions.java:
```diff
@@ -40,4 +53,83 @@ public enum Stack {
+            .buildTime(true)
+            .build();
+
+    public static final Option<String> JGROUPS_TLS_CLIENT_AUTH = new OptionBuilder<>(JGROUPS_TLS_CLIENT_AUTH_PROPERTY, String.class)
```
From what we discussed earlier we want to prevent unauthorized clients to join the communication. So the only option that makes sense IMHO is `need` here. If anything else is not a common use case, we should remove this option and hard-code it to `need`. A user of Keycloak can still change the XML if they want to do something similar. 

If we allow for this option to be changed, I assume people will choose an unsecure setting by accident.

pruivo replied:
Makes sense.

ahus1 commented on quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:
```diff
@@ -129,8 +140,46 @@ private void shutdownThreadPool() {
+            transportConfig.defaultTransport().stack(transportStack);
         }
+
+        if (booleanProperty(JGROUPS_TLS_ENABLED_PROPERTY)) {
```
Would it be correct to assume that TLS works only with TCP as a transport portocol? If yes, please add a check for that, and throw an exception if a mismatch is detected.

pruivo replied:
Yes, this only works with TCP sockets. UDP sockets or Java NIO channels are not supported.

ahus1 replied:
I saw your latest changes for this. I suppose it works, still I wonder why this is doing a several negative checks (things that shouldn't be present) instead of a positive check that it is using TCP.

ahus1 reviewed (changes requested):
Thank you for this PR - I'm looking forward to see the communication finally encrypted :-)

The documentation from a CLI perspective should probably go here https://www.keycloak.org/server/caching#_securing_cache_communication. Adding it to the list of things to do in https://www.keycloak.org/server/configuration-production would also be good.

I'm not sure we then need additional documentation for the Operator. Once we've settled on the docs and the options, I'd like to see them in the HA guide eventually.

I see the options are currently named `jgroups-tls-...`. As users of Keycloak know on "the cache", I'd suggest to rename it `cache-tls-...` and hide JGroups as an implementation detail.

I still want to test this manually in a cluster, and we would like to see a test case for this in Keycloak before we can merge this.

ahus1 commented:
Hi @keycloak/cloud-native team - with @pruivo knowledge about Infinispan, we would like offer Keycloak CLI options to enable encrypted JGroup communication used for Infinispan between nodes. 

In a nutshell: create a self-signed certificate, add it to the CLI to use it for all internal JGroups communications of the embedded cache, and be safe.

A thing to watch out for: It need be a self-signed certificate used only for this purpose, as only those who have a client certificate signed with that CA would should be allowed to connect. The good thing: no other permissions need to configured for this. It must not use a CA used also for other purposes, as that would allow a lot of other clients to connect as well. 

We'd like to get your early feedback on this PR, for example on the naming of the CLI options. We know that docs and a test are still missing. 

I'm not sure if we would like to see a test Keycloak-only, or one that is using the Operator.

ahus1 commented:
Update: Manual test where I deployed this on our OpenShift cluster worked nicely.

pruivo commented:
Updated
* Added some docs that require some love
* Renamed `jgroups-tls` with `cache-tls`
* Added validation to check if the protocol used supports TLS or not

ahus1 commented:
Agreed namings with @vmuzikar in today's call + reduced list to avoid too many configuration options: 

```
    - name: cache-embedded-mtls-enabled
      value: "true"
    - name: cache-embedded-mtls-key-store-file
      value: /etc/jgroups-tls/keystore.p12
    - name: cache-embedded-mtls-key-store-password
      secret:
        name: keycloak-jgroups-tls-secret
        key: keystorePassword
    - name: cache-embedded-mtls-key-store-alias
      # HTTP server works without a key alias, can we get rid of the option? 
      value: "site1"
    - name: cache-embedded-mtls-trust-store-file
      value: /etc/jgroups-tls/truststore.p12
    - name: cache-embedded-mtls-trust-store-password
      secret:
        name: keycloak-cache-tls-secret
        key: truststorePassword
```

In the future, the Operator could possibly provide the truststore and keystore without the need for the user to provide them.

sschu commented:
Ideally using https://cert-manager.io/.

ahus1 commented:
@sschu - IMHO this doesn't work here, as we need a private CA per KC instance for this. Using a public CA would allow all certificates issued by that CA access to the Keycloak Infinispan setup, as the setup would accept any client certificate signed by the CA.

ahus1 commented:
@pruivo - switching the internal communication to TLS, would it be affected by the bugs discovered around https://github.com/keycloak/keycloak-benchmark/issues/675? 

That is: Once we switch the internal JGroups communication to TLS, it might stall in a similar way if a Keycloak server is not responding, or would it behave better due to the failure detection mechanism?

pruivo commented:
@ahus1 I would have to test it. Will do that next Monday since I have a doctor appointment this afternoon.

ahus1 commented:
Sure, no hurry. Just a thought and I'm curious. Let's solve the Gossip router situation first.

sschu commented:
@ahus1 You can use multiple CA's with cert-manager, if I am not mistaken: https://cert-manager.io/docs/configuration/ca/

ahus1 commented:
@pruivo - I've updated the parameters to match the suggestions by @vmuzikar, by adding additional `-`, and removed those were he would to use the default values always.

One parameter remains that should be removed: Can we get rid of the keystore alias? Would it "just work" without? In all other examples (for example the key for HTTPS) we don't pass it, so we would like to get rid of that parameter as well.

pruivo commented:
@ahus1 I removed the alias. Tested with a self-signed keystore that I use for testing and with a keystore generated from cert-manager.

ahus1 committed:
Review for encrypting network communication in JGroups

Closes #25702

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 commented:
@pruivo - I reviewed the PR and cleaned up the properties. 

When testing it, I found it would be helpful to add a log message to say that MTLS is now enabled, as mis-typing the parameters wouldn't throw any error when using the operator. I also found that the 4 parameters for keystore/truststore should be required, again for avoiding mistyped parameters. I found out the hard way that it would use the default keystore/truststore in that case, and that would lead most likely to a misconfiguration as this is supposed to be used with a self-signed certificate in our case. 

I also rebased it and updated the help files to the latest state.

I tested this with PR https://github.com/keycloak/keycloak-benchmark/pull/685 on KCB and OpenShift with a locally built Keycloak image.

As I did some changes, I'd like you to have a look at my commit. If you agree, please comment here, or add another commit with any changes you'd like to see.

As the CLI options now align with what we agreed with @vmuzikar, I assume it is fine to be merged afterwards. 

Thanks!

ahus1 reviewed (approved):
Worked in my tests after the review changes. Waiting for review of my changes by @pruivo

pruivo commented:
@ahus1 is looks good to me :+1:

vmuzikar commented:
@pruivo @ahus1 Thanks for the changes! I plan to review tomorrow, or early next week.
